### ISSUE TITLE
Data Annotations in Nested Data Models

### ISSUE THREAD

**Issue by TiriTM** [cofoundry-cms/cofoundry#342]
Hi there I'm currently stucked with a problem I don't understand why it's happening. We're using the **Cofoundry Web Admin 0.6.1 NuGet Package** which has been published on April 13th 2019.

The following is a code snippet from our nested DataModel of a product slider we're building for a book webshop.

```
public class ProductPickerSlideDataModel : INestedDataModel
{
	[PreviewTitle]
	[Required(ErrorMessageResourceName = "Required_ISBN", ErrorMessageResourceType = typeof(Localization.Web.CMS.Global))]
	[RegularExpression(@"^(?:97[89][- ]?)?[0-9]{1,5}[- ]?[0-9]+[- ]?[0-9]+[- ]?[0-9X]$", ErrorMessageResourceName = "MalformedISBN", ErrorMessageResourceType = typeof(Localization.Web.CMS.ProductPicker))]
	[Display(Name = "Suche", Description = "Geben Sie eine ISBN oder eine ISSN ein, um nach einem Titel zu suchen.")]
	public string Query { get; set; }
}
```

Currently the `Required` annotation works **partially** here, as the field is marked as required correctly and the message "This field is required" does appear if I'd leave it blank. But in fact, it should show another message defined in our localization resource. The `RegularExpression` annotation doesn't work at all, so I could enter "Hitchhiker's" instead of an ISBN and it gets recognized as valid while in fact it's not. Does anyone know some help here? What am I doin' wrong?

Kind regards and thanks in advance.

**Comment by HeyJoel** [comment]
So there's two issues here:

1) The regular expression annotation isn't working because usually this is done server-side but your case your using a nested data model which is not processed server-side. So this is a bug, the solution for us will be to run server-side validation before adding the value.
2) Localization: I'm not overly familiar with resx files in .NET Core, other than knowing that they changed a lot of stuff and it's a bit half-baked. I can't see any reference to `ErrorMessageResourceType` in the [docs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-2.2), is it still used? Do check that what you're trying to do works outside of Cofoundry first, if it's a Cofoundry issue then we'd probably look to address it as part of #43 (localization).

**Comment by TiriTM** [comment]
@HeyJoel thanks for your feedback, in fact "ErrorMessageResourceType" works well outside of Cofoundry and even **within** Cofoundry, but especially **not** in the Nested Data Model. We're successfully using this in a call2action control we've created and where we check a url input field against valid urls OR html-anchors using `#id`-Syntax.

```
[RegularExpression(@"^(?:(ht|f)tps?:\/{2})?[\w#.-]+(?:\.[\w\.-]+)*[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$", ErrorMessageResourceName = "MalformedLinkTarget", ErrorMessageResourceType = typeof(Localization.Web.CMS.CallToAction))]
[Display(Name = "Ziel", Description = "Optional, die URL des Call2Action-Buttons, ohne diese URL wird der Button/Link nicht angezeigt.")]
[MaxLength(256)]
public string Url { get; set; }

```
This works like a charme, our custom validation message is displayed if I try to enter something that's neither a valid url nor something like `#id`. We did this because the [Url] annotation led to a foced `http://` at the beginning of anything I enter in this field. We needed to be able to enter something like `#novelties` or '#bestsellers' to make maintenance of internal links easier for our customer, so we decided to use such a regex with a custom validation message.

As this (not working of ErrorMessageResourceType in Nested Data Model) seems to be a bug as well I will have to look for another way to achieve what we need here (for this time) thanks anyway :)

**Comment by HeyJoel** [comment]
Thanks for the info, I imagine the

**Comment by HeyJoel** [comment]
Thanks for the info, .NET must be doing something specific on postback to resolve those validation messages so I'll have to dig into that and see if we can replicate it with our model schema builder.

If you wanted to look into the implementation [`DynamicDataModelSchemaMapper`](https://github.com/cofoundry-cms/cofoundry/blob/22507ed1946aaf279df59590b81a963d14b50dbc/src/Cofoundry.Domain/Domain/DynamicData/Mapping/DynamicDataModelSchemaMapper.cs) is a good place to start.

**Comment by HeyJoel** [comment]
Server-side validation now runs before adding/updating a nested data model. This will be released in 0.8.1.
